### PR TITLE
Fix race condition allowing multiple Tick() calls to a single Shard

### PR DESCRIPTION
**Pull request by prateek**
/cc @robskillington @richardartoul

### CODE REVIEW

File: storage/shard.go [hunk 1]
```diff
@@ -354,27 +361,61 @@ func (s *dbShard) Close() error {
...
-	return s.tickAndExpire(c, softDeadline, tickPolicyRegular)
+	return s.tickAndExpire(c, softDeadline, tickPolicyRegular, defaultSkipTickIfShardClosing)
 }
 
 func (s *dbShard) tickAndExpire(
 	c context.Cancellable,
 	softDeadline time.Duration,
 	policy tickPolicy,
-) tickResult {
+	skipIfClosing bool,
```
Comment (robskillington) [inline storage/shard.go]: Can you combine `tickPolicy` and `skipIfClosing`?  i.e. just change `tickPolicyForceExpiry` to `tickPolicyCloseShard`

File: storage/shard.go [hunk 2]
```diff
@@ -354,27 +361,61 @@ func (s *dbShard) Close() error {
...
+) (tickResult, error) {
+	// ensure only one tick can execute at a time
+	if s.ticking.Swap(true) {
+		// i.e. we were previously ticking
+		return tickResult{}, errShardAlreadyTicking
+	}
```
Comment (robskillington) [inline storage/shard.go]: If the tickPolicy is `tickPolicyRegular` and `s.isClosing()` can we also return err?

File: storage/shard.go [hunk 3]
```diff
@@ -354,27 +361,61 @@ func (s *dbShard) Close() error {
...
+	s.tickWg.Wait()
+
 	// NB(r): Asynchronously we purge expired series to ensure pressure on the
 	// GC is not placed all at one time.  If the deadline is too low and still
 	// causes the GC to impact performance when closing shards the deadline
 	// should be increased.
 	cancellable := context.NewNoOpCanncellable()
 	softDeadline := s.opts.TickInterval()
-	s.tickAndExpire(cancellable, softDeadline, tickPolicyForceExpiry)
+	s.tickAndExpire(cancellable, softDeadline, tickPolicyForceExpiry, false)
```
Comment (robskillington) [inline storage/shard.go]: Probably want to check if this returns an error and return it yeah?

File: storage/shard.go (hunk 1 above)
Comment (prateek) [reply storage/shard.go]: sure thing

File: storage/shard.go (hunk 2 above)
Comment (prateek) [reply storage/shard.go]: sure thing

File: storage/shard.go (hunk 3 above)
Comment (prateek) [reply storage/shard.go]: nice catch, will do

Comment (prateek) [reply storage/shard.go]: done

File: storage/shard.go (hunk 2 above)
Comment (prateek) [reply storage/shard.go]: done

File: storage/shard.go (hunk 1 above)
Comment (prateek) [reply storage/shard.go]: done

File: storage/namespace.go [hunk 4]
```diff
@@ -435,6 +438,8 @@ func (n *dbNamespace) Tick(c context.Cancellable) {
 	n.metrics.tick.madeUnwiredBlocks.Inc(int64(r.madeUnwiredBlocks))
 	n.metrics.tick.mergedOutOfOrderBlocks.Inc(int64(r.mergedOutOfOrderBlocks))
 	n.metrics.tick.errors.Inc(int64(r.errors))
+
+	return multiErr.FinalError()
```
Comment (robskillington) [inline storage/namespace.go]: I don't think we want to emit stats or set the last tick stats unless err == nil.

Comment (prateek) [reply storage/namespace.go]: fixed up

File: glide.yaml [hunk 5]
```diff
@@ -89,6 +89,9 @@ import:
 - package: github.com/spf13/pflag
   version: 4f9190456aed1c2113ca51ea9b89219747458dc1
 
+- package: go.uber.org/atomic
```
Comment (robskillington) [inline glide.yaml]: Do we need to pull this in anymore?

File: storage/namespace.go [hunk 6]
```diff
@@ -406,18 +407,24 @@ func (n *dbNamespace) Tick(c context.Cancellable) {
...
+			multiErr = multiErr.Add(err)
 			l.Unlock()
 		})
 	}
 
 	wg.Wait()
 
 	if c.IsCancelled() {
-		return
+		return multiErr.FinalError()
```
Comment (robskillington) [inline storage/namespace.go]: Is it worth actually returning here?  As long as err == nil even if it just got cancelled we should probably inc stats, etc.

File: glide.yaml (hunk 5 above)
Comment (prateek) [reply glide.yaml]: doh I remember thinking ill forget to revert that :|

File: storage/namespace.go [hunk 7]
```diff
@@ -406,18 +407,24 @@ func (n *dbNamespace) Tick(c context.Cancellable) {
...
+	}
+
+	if err := multiErr.FinalError(); err != nil {
```
Comment (robskillington) [inline storage/namespace.go]: Maybe combine these two together? e.g. `if err := multiErr.FinalError(); err != nil || c.IsCancelled() { return err }`

File: storage/shard_test.go [hunk 8]
```diff
@@ -476,6 +487,145 @@ func TestShardWriteAsync(t *testing.T) {
 	require.True(t, ok)
 }
 
+// This tests a race in shard ticking with an empty series pending expiration.
+func TestShardTickRace(t *testing.T) {
```
Comment (robskillington) [inline storage/shard_test.go]: Good set of tests 👍

File: storage/namespace.go (hunk 7 above)
Comment (prateek) [reply storage/namespace.go]: sure thing

Comment (prateek) [reply storage/namespace.go]: done

**Pull request merged**